src/local/experiment.py

```diff
@@ -72,29 +75,32 @@ def load_data(self):
         
     def load_sample_indices(self):
         self.logger.info("Loading sample indices")
+        test_samples_limit = self.cfg.experiment.get('test_samples_limit', None)
         self.sample_indices_train = get_sample_indices(self.cfg.data.genotype.train,
                                                        self.cfg.data.phenotype.train)
         self.sample_indices_val = get_sample_indices(self.cfg.data.genotype.val,
                                                      self.cfg.data.phenotype.val)
         self.sample_indices_test = get_sample_indices(self.cfg.data.genotype.test,
-                                                      self.cfg.data.phenotype.test)
+                                                      self.cfg.data.phenotype.test,
+                                                      indices_limit=test_samples_limit)
         
     def load_genotype_and_covariates_(self):
+        test_samples_limit = self.cfg.experiment.get('test_samples_limit', None)
         self.X_train = hstack((load_from_pgen(self.cfg.data.genotype.train,
                                               self.cfg.data.gwas,
                                               snp_count=self.cfg.experiment.snp_count,
                                               sample_indices=self.sample_indices_train),
-                               load_covariates(self.cfg.data.covariates.train)))
+                               load_covariates(self.cfg.data.covariates.train).astype(numpy.float16)))
```

Comment:
just asking, is there really much memory benefit from having this as float16? Is it 4 digits? I'm not sure that's enough for, say, PCs

Reply:
hstack automatically casts arrays into their largest type. If covariates are in float32, then whole genotype matrix will also be in float32. We have two choices here: accept loss of precision for covariates with values around zero or rewrite data loading to store genotype and covariate matrices in different numpy arrays

Reply:
The whole genotype matrix takes 600k*1.5k*2=1.8GB in float16 for the smallest node. For the largest node it will take 600k*200k*2=240GB. This is already too close to 360GB limit on cluster GPU nodes